extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js

```diff
@@ -258,6 +423,117 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
   }
 
   /** @override */
+  sendXhrRequest(adUrl) {
+    if (!this.useSra_) {
+      return super.sendXhrRequest(adUrl);
+    }
+    // Wait for SRA request which will ncall response promise when this block's
+    // response has been returned.
+    this.initiateSraRequests_();
+    // Null response indicates single slot should execute using non-SRA method.
+    return this.sraResponsePromise_.then(
+      response => response || super.sendXhrRequest(adUrl));
+  }
+
+  /**
+   * Executes SRA request via the following steps:
+   * - create only one executor per page
+   * - get all doubleclick amp-ad instances on the page
+   * - group by networkID allowing for separate SRA requests
+   * - for each grouping, construct SRA request
+   * - handle chunks for streaming response for each block
+   * @private
+   */
+  initiateSraRequests_() {
+    if (SraRequests) {
+      return;
+    }
+    // Use cancellation of the first slot's promiseId as indication of
+    // unlayoutCallback execution.  Assume that if called for one slot, it will
+    // be called for all and we should cancel SRA execution.
+    const checkStillCurrent = this.verifyStillCurrent();
+    SraRequests = groupAmpAdsByType(
+        this.win, this.element.getAttribute('type'), getNetworkId_)
+      .then(groupIdToBlocksAry => {
+        checkStillCurrent();
+        Object.keys(groupIdToBlocksAry).forEach(networkId => {
+          const blocks = dev().assert(groupIdToBlocksAry[networkId]);
+          // TODO: filter blocks with SRA disabled?
+          return Promise.all(blocks).then(instances => {
+            dev().assert(instances.length);
+            const typeInstances =
+              /** @type {!Array<!AmpAdNetworkDoubleclickImpl>}*/(instances);
+            // Determine if more than one block for this element, if not do not
+            // set sra request promise which results in sending as
+            // non-SRA request (benefit is it allows direct cache method).
+            if (typeInstances.length == 1) {
+              dev().info(TAG, `single block in network ${networkId}`);
+              typeInstances[0].sraResponseResolver_(null);
```

Comment:
Yes and no. It'll work, because Closure doesn't enforce private. But private properties are only for _this_ instance, you can't touch another instances private properties.